horovod/tensorflow/__init__.py

```diff
@@ -477,6 +479,10 @@ def DistributedOptimizer(optimizer, name=None, use_locking=False, device_dense='
         before and after the sum. Gradients are scaled by
         1.0 / gradient_predivide_factor before the sum and
         gradient_predivide_factor / size after the sum.
+      average_aggregated_gradients:
```

Comment:
We have seen some use cases where users prefer to manage these reduction via other techniques such as loss scaling but I do agree that if gradients are being averaged likely they want to average both across workers and across aggregations. So if we want to avoid adding an additional argument I am open to just using `op` to set this behavior.

Reply:
I think just using `op` is fine to avoid confusion.  If we kept this parameter, we would also need to think about how it applies to Adasum.  For now, if the user wants to do custom reductions, they should be able to set `op=Sum` and then handle it elsewhere.

Reply:
If we are switching over to using `op` to toggle averaging aggregated gradients, should I also add `op` as an argument for the [Keras optimizer](https://github.com/horovod/horovod/blob/56183ca42c43aad7afd619f0cc8bc4842336f3ec/horovod/tensorflow/keras/__init__.py#L48) to make it configurable.

Reply:
One other factor worth considering here is that if we switch over to using `op` to control averaging across `backewards_passes_per_step` this would be inconsistent with the current behavior of the PyTorch[ _DistributedOptimier](https://github.com/horovod/horovod/blob/56183ca42c43aad7afd619f0cc8bc4842336f3ec/horovod/torch/optimizer.py#L32).

Reply:
Good catch.

Regarding Keras: I think it was an oversight that we don't expose `op` as a param to the Keras optimizer, so making that change sounds reasonable to me.

Regarding PyTorch: that's a very a good point.  Currently, this is handled in user code [here](https://github.com/horovod/horovod/blob/master/examples/pytorch/pytorch_imagenet_resnet50.py#L83).

Looking at Adasum, it looks like they don't do any averaging across the local aggregations either.

I think keeping `average_aggregated_gradients` may be the best way to go. However, we may want to set it to `False` by default, and raise an exception for now when the user sets it to `True` along with `op=Adasum`.  Otherwise, the user will get inconsistent behavior between `Average` and `Adasum` when setting `backward_passes_per_step > 1`.  Would that work for your users?

Reply:
Yeah this approach makes a lot of sense and allows the most consistency across the different optimizer types. I will go ahead and add `op` as a configuration to the Keras DistributedOptimizer, set the default for `average_aggregated_gradients` to `False` and add a check for  when `op=Adasum` to make sure `average_aggregated_gradients` is set to False. I think down the road it probably makes sense to add `average_aggregated_gradients` to the PyTorch optimizer as well.

Reply:
I agree, adding it as a feature for other optimizers / modes would be a great addition in the future.  Thanks for digging into this!